pennylane_qiskit/qiskit_device2.py

```diff
@@ -0,0 +1,625 @@
+
+        overlapping_kwargs = set(self._init_kwargs).intersection(set(option_kwargs))
+        if overlapping_kwargs:
+            warnings.warn(
+                f"The keyword argument(s) {overlapping_kwargs} passed to the device are also "
+                f"defined in the device Options. The definition in Options will be used."
+            )
+        if option_kwargs["shots"] != self.shots.total_shots:
+            warnings.warn(
+                f"Setting shots via the Options is not supported on PennyLane devices. The shots {self.shots} "
+                f"passed to the device will be used."
+            )
+            self.options.execution.shots = self.shots.total_shots
+
+        option_kwargs.pop("shots")
+        kwargs = self._init_kwargs.copy()
+        kwargs.update(option_kwargs)
+
+        self._kwargs = kwargs
+
+    @staticmethod
+    def get_transpile_args(kwargs):
+        """The transpile argument setter.
+
+        Keyword Args:
+            kwargs (dict): keyword arguments to be set for the Qiskit transpiler. For more details, see the
+                `Qiskit transpiler documentation <https://qiskit.org/documentation/stubs/qiskit.compiler.transpile.html>`_
+        """
+
+        transpile_sig = inspect.signature(transpile).parameters
+
+        transpile_args = {arg: kwargs[arg] for arg in transpile_sig if arg in kwargs}
+        transpile_args.pop("circuits", None)
+        transpile_args.pop("backend", None)
+
+        return transpile_args
+
+    def compile_circuits(self, circuits):
+        r"""Compiles multiple circuits one after the other.
+
+        Args:
+            circuits (list[QuantumCircuit]): the circuits to be compiled
+
+        Returns:
+             list[QuantumCircuit]: the list of compiled circuits
+        """
+        # Compile each circuit object
+        compiled_circuits = []
+        transpile_args = self.get_transpile_args(self._kwargs)
+
+        for i, circuit in enumerate(circuits):
+            compiled_circ = transpile(circuit, backend=self.backend, **transpile_args)
+            compiled_circ.name = f"circ{i}"
+            compiled_circuits.append(compiled_circ)
+
+        return compiled_circuits
+
+    # pylint: disable=unused-argument
+    def execute(
+        self,
+        circuits: QuantumTape_or_Batch,
+        execution_config: ExecutionConfig = DefaultExecutionConfig,
+    ) -> Result_or_ResultBatch:
+        session = self._session or Session(backend=self.backend)
+
+        if not self._use_primitives:
+            results = self._execute_runtime_service(circuits, session=session)
+            return results
+
+        results = []
+
+        if isinstance(circuits, QuantumScript):
+            circuits = [circuits]
+
+        @contextmanager
+        def execute_circuits(session):
+            try:
+                for circ in circuits:
+                    if circ.shots:
+                        self.options.execution.shots = circ.shots.total_shots
```

Comment:
> Also a valid input in Pennylane would be `circuit(shots=[2, 10, 5])`, currently this would give 3 results with differing shot numbers in, for example, `default.qubit`.
> 
> That information is lost by using `.total_shots`, since those would be the sum of all the shots, in this case 17. I think that's a fine way to handle this for the Qiskit device, but maybe we should raise a warning in the case where `len(shots.shot_vector > 1)`, since we aren't giving people what they ask for in that case.

Ah, I totally didn't realize that that was a possible case in Pennylane. I'll raise a warning in this case and let the user know that that that really isn't supported by Qiskit and what was actually run.

Reply:
Addressed the shot tracking behaviour, added tests and also warning for shot vector. For testing the shots I did a bit of a hack by saving the result that comes from the IBM devices directly to device._current_job and then pulling the shots information from there to verify. 

e.g.
`assert dev._current_job.metadata[0]["shots"] == 5`

Could also make a getter function that gets us the shot information instead, not sure if that's worth the effort though. If you have any preference, let me know and I'll write a getter function and call that instead.